File: src/core/cors.ts
// Canonical Core — CORS: the one and only trust primitive.
//
// Trust is NOT based on origin. A resource is reachable only when its provider
// has issued an explicit, signed sharing contract. Only CORS is trusted.
//
// A grant IS a contract. Only the provider — the resource owner — holds the
// authority to originate one. Every valid grant chains, by monotone
// attenuation, back to that provider; a consumer only ever holds a delegated
// sub-contract, never authority of its own.

import { signRecord, verifyRecord, type Signer } from "./identity.ts";
import { actionSubsumes, resourceSubsumes } from "./resource.ts";

export interface Caveat {
  readonly type: string;
  readonly [k: string]: unknown;
}

export interface CorsGrant {
  /** who signed this link of the contract — the provider, or a re-delegator. */
  readonly issuer: string;
  /** who may use it. */
  readonly subject: string;
  /** the action, e.g. "invoke" / "read"; "*" means any. */
  readonly action: string;
  /** the resource "<ownerId>:<path>"; path may end with "/*". */
  readonly resource: string;
  readonly caveats?: readonly Caveat[];
  readonly issuedAt: number;
  /** parent contract, embedded so the edge can verify with no external lookup. */
  readonly proof?: CorsGrant;
  readonly sig: string;
}

/** Does this contract, on its face, permit `action` on `resource`? */
export function authorizes(grant: CorsGrant, action: string, resource: string): boolean {
  return actionSubsumes(grant.action, action) && resourceSubsumes(grant.resource, resource);
}

function caveatsHold(caveats: readonly Caveat[] | undefined, now: number): GrantCheck {
  for (const c of caveats ?? []) {
    if (c.type === "expiry") {
      const notAfter = c["notAfter"];
      if (typeof notAfter !== "number" || now > notAfter) return { ok: false, reason: "contract expired" };
    } else {
      // Fail closed: an edge that does not understand a caveat must not honour it.
      return { ok: false, reason: `unknown caveat '${c.type}'` };
    }
  }
  return { ok: true };
}

export interface IssueParams {
  subject: string;
  action: string;
  resource: string;
  caveats?: readonly Caveat[];
  proof?: CorsGrant;
}

/** The provider originates a contract for one of its own resources. */
export async function issueGrant(owner: Signer, params: IssueParams, now: number = Date.now()): Promise<CorsGrant> {
  const body = {
    issuer: owner.id,
    subject: params.subject,
    action: params.action,
    resource: params.resource,
    caveats: params.caveats,
    issuedAt: now,
    proof: params.proof,
  };
  return signRecord(owner, body);
}

export interface AttenuateParams {
  subject: string;
  action?: string;
  resource?: string;
  caveats?: readonly Caveat[];
}

/** A holder delegates part of a contract it holds to another subject. */
export async function attenuate(
  holder: Signer,
  parent: CorsGrant,
  params: AttenuateParams,
  now: number = Date.now(),
): Promise<CorsGrant> {
  return issueGrant(
    holder,
    {
      subject: params.subject,
      action: params.action ?? parent.action,
      resource: params.resource ?? parent.resource,
      caveats: params.caveats,
      proof: parent,
    },
    now,
  );
}

export interface GrantCheck {
  readonly ok: boolean;
  readonly reason?: string;
}

// Verify the contract chains, by valid signatures and monotone attenuation, all
// the way back to the resource provider. There is no origin shortcut: only the
// provider can root a contract, and no link may widen what its parent granted.
export async function verifyGrant(grant: CorsGrant, providerId: string, now: number, depth = 0): Promise<GrantCheck> {
  if (depth > 32) return { ok: false, reason: "contract chain too deep" };
  if (!(await verifyRecord(grant.issuer, grant))) return { ok: false, reason: "invalid contract signature" };
  const cav = caveatsHold(grant.caveats, now);
  if (!cav.ok) return cav;
  if (grant.issuer === providerId) return { ok: true };
  if (!grant.proof) return { ok: false, reason: "contract not rooted at the provider" };
  if (grant.proof.subject !== grant.issuer) return { ok: false, reason: "delegation not held by its issuer" };
  if (!actionSubsumes(grant.proof.action, grant.action)) return { ok: false, reason: "delegation widens the action" };
  if (!resourceSubsumes(grant.proof.resource, grant.resource)) return { ok: false, reason: "delegation widens the resource" };
  return verifyGrant(grant.proof, providerId, now, depth + 1);
}

File: crates/canonical-core/src/cors.rs
//! CORS: the one and only trust primitive.
//!
//! Trust is not based on origin. A resource is reachable only when its provider
//! has issued an explicit, signed sharing contract. A grant IS a contract: only
//! the provider can originate one, and every valid grant chains, by monotone
//! attenuation, back to that provider.

use std::collections::BTreeMap;

use crate::codec::Json;
use crate::identity::{verify_record, Signer};
use crate::resource::{action_subsumes, resource_subsumes};

#[derive(Clone, Debug)]
pub struct Caveat {
    pub kind: String,
    pub not_after: Option<i64>,
    pub not_before: Option<i64>,
}

impl Caveat {
    /// A short-lived contract: invalid after `not_after` (ms since epoch).
    pub fn expiry(not_after: i64) -> Caveat {
        Caveat {
            kind: "expiry".into(),
            not_after: Some(not_after),
            not_before: None,
        }
    }

    /// A contract that only becomes valid at `not_before`.
    pub fn not_before(not_before: i64) -> Caveat {
        Caveat {
            kind: "notBefore".into(),
            not_after: None,
            not_before: Some(not_before),
        }
    }

    pub fn unknown(kind: impl Into<String>) -> Caveat {
        Caveat {
            kind: kind.into(),
            not_after: None,
            not_before: None,
        }
    }

    fn json(&self) -> Json {
        let mut m = BTreeMap::new();
        m.insert("type".to_string(), Json::str(self.kind.clone()));
        if let Some(n) = self.not_after {
            m.insert("notAfter".to_string(), Json::Int(n));
        }
        if let Some(n) = self.not_before {
            m.insert("notBefore".to_string(), Json::Int(n));
        }
        Json::Obj(m)
    }
}

#[derive(Clone, Debug)]
pub struct CorsGrant {
    pub issuer: String,
    pub subject: String,
    pub action: String,
    pub resource: String,
    pub caveats: Vec<Caveat>,
    pub issued_at: i64,
    pub proof: Option<Box<CorsGrant>>,
    pub sig: String,
}

impl CorsGrant {
    /// The signed body (everything but `sig`). An embedded proof carries its own
    /// signature, so the contract is self-contained and verifiable at the edge.
    pub fn signing_json(&self) -> Json {
        let mut m = BTreeMap::new();
        m.insert("action".to_string(), Json::str(self.action.clone()));
        if !self.caveats.is_empty() {
            m.insert(
                "caveats".to_string(),
                Json::Arr(self.caveats.iter().map(Caveat::json).collect()),
            );
        }
        m.insert("issuedAt".to_string(), Json::Int(self.issued_at));
        m.insert("issuer".to_string(), Json::str(self.issuer.clone()));
        if let Some(p) = &self.proof {
            m.insert("proof".to_string(), p.full_json());
        }
        m.insert("resource".to_string(), Json::str(self.resource.clone()));
        m.insert("subject".to_string(), Json::str(self.subject.clone()));
        Json::Obj(m)
    }

    fn full_json(&self) -> Json {
        let mut j = self.signing_json();
        if let Json::Obj(m) = &mut j {
            m.insert("sig".to_string(), Json::str(self.sig.clone()));
        }
        j
    }
}

/// Does this contract, on its face, permit `action` on `resource`?
pub fn authorizes(grant: &CorsGrant, action: &str, resource: &str) -> bool {
    action_subsumes(&grant.action, action) && resource_subsumes(&grant.resource, resource)
}

pub struct IssueParams<'a> {
    pub subject: &'a str,
    pub action: &'a str,
    pub resource: &'a str,
    pub caveats: Vec<Caveat>,
    pub proof: Option<CorsGrant>,
}

/// The provider originates a contract for one of its own resources.
pub fn issue_grant(owner: &Signer, params: IssueParams, now: i64) -> CorsGrant {
    let mut g = CorsGrant {
        issuer: owner.id.clone(),
        subject: params.subject.to_string(),
        action: params.action.to_string(),
        resource: params.resource.to_string(),
        caveats: params.caveats,
        issued_at: now,
        proof: params.proof.map(Box::new),
        sig: String::new(),
    };
    g.sig = owner.sign_record(&g.signing_json());
    g
}

/// A holder delegates part of a contract it holds to another subject.
pub fn attenuate(
    holder: &Signer,
    parent: &CorsGrant,
    subject: &str,
    action: Option<&str>,
    resource: Option<&str>,
    caveats: Vec<Caveat>,
    now: i64,
) -> CorsGrant {
    issue_grant(
        holder,
        IssueParams {
            subject,
            action: action.unwrap_or(&parent.action),
            resource: resource.unwrap_or(&parent.resource),
            caveats,
            proof: Some(parent.clone()),
        },
        now,
    )
}

#[derive(Debug)]
pub struct GrantCheck {
    pub ok: bool,
    pub reason: Option<String>,
}

fn deny(reason: &str) -> GrantCheck {
    GrantCheck {
        ok: false,
        reason: Some(reason.to_string()),
    }
}
fn ok() -> GrantCheck {
    GrantCheck {
        ok: true,
        reason: None,
    }
}

/// Verify the contract chains, by valid signatures and monotone attenuation, all
/// the way back to the resource provider. Only the provider can root a contract,
/// and no link may widen what its parent granted.
pub fn verify_grant(grant: &CorsGrant, provider_id: &str, now: i64) -> GrantCheck {
    verify_grant_depth(grant, provider_id, now, 0)
}

fn verify_grant_depth(grant: &CorsGrant, provider_id: &str, now: i64, depth: u32) -> GrantCheck {
    if depth > 32 {
        return deny("contract chain too deep");
    }
    if !verify_record(&grant.issuer, &grant.signing_json(), &grant.sig) {
        return deny("invalid contract signature");
    }
    for c in &grant.caveats {
        match c.kind.as_str() {
            "expiry" => match c.not_after {
                Some(n) if now <= n => {}
                _ => return deny("contract expired"),
            },
            "notBefore" => match c.not_before {
                Some(n) if now >= n => {}
                _ => return deny("contract not yet valid"),
            },
            // fail closed: an edge that does not understand a caveat must not honour it
            other => return deny(&format!("unknown caveat '{other}'")),
        }
    }
    if grant.issuer == provider_id {
        return ok();
    }
    let proof = match &grant.proof {
        Some(p) => p,
        None => return deny("contract not rooted at the provider"),
    };
    if proof.subject != grant.issuer {
        return deny("delegation not held by its issuer");
    }
    if !action_subsumes(&proof.action, &grant.action) {
        return deny("delegation widens the action");
    }
    if !resource_subsumes(&proof.resource, &grant.resource) {
        return deny("delegation widens the resource");
    }
    verify_grant_depth(proof, provider_id, now, depth + 1)
}
